### PR TITLE
Tsubomi/auditlog

### PR DESCRIPTION
**Pull request by leitsubomi**
Add emit log object in audit.js with unit test
Add unit test to check current route duplication behavior

### CODE REVIEW

**Comment by micahr** [comment]
It's a little tough to see what changes were made due to d993d6a is it possible for that to be in a separate PR?

**Comment by yunong** [comment]
Can you remove 146ab0c and make it its own PR?

File: lib/plugins/audit.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ function auditLogger(options) {
                 return ({
                     // account for native and queryParser plugin usage
                     query: (typeof req.query === 'function') ?
-                            req.query() : req.query,
+                        req.query() : req.query,
```
Comment (yunong) [inline lib/plugins/audit.js]: Please revert this line.

File: lib/plugins/audit.js [hunk 2]
```diff
@@ -91,6 +91,7 @@ function auditLogger(options) {
 
     function audit(req, res, route, err) {
         var latency = res.get('Response-Time');
+        var server = options.server;
```
Comment (yunong) [inline lib/plugins/audit.js]: Cache the server up top when the audit function is initialized in its own variable.

File: lib/plugins/audit.js [hunk 3]
```diff
@@ -25,6 +24,7 @@ var HttpError = require('restify-errors').HttpError;
 function auditLogger(options) {
```
Comment (yunong) [inline lib/plugins/audit.js]: Add docs about the server object. You'll also want to optionally assert that the server parameter is an object.

File: test/plugins.test.js [hunk 4]
```diff
@@ -76,17 +78,13 @@ after(function (callback) {
         for (i = 0; i < FILES_TO_DELETE.length; ++i) {
             try {
                 fs.unlinkSync(FILES_TO_DELETE[i]);
-            }
-            catch (err) { /* normal */
-            }
+            } catch (err) { /* normal */ }
```
Comment (yunong) [inline test/plugins.test.js]: Revert this change.

File: test/plugins.test.js [hunk 5]
```diff
@@ -232,7 +230,9 @@ test('body url-encoded ok', function (t) {
 
 test('body url-encoded ok (no params)', function (t) {
     SERVER.post('/bodyurl2/:id',
-        restify.bodyParser({ mapParams: false }),
```
Comment (yunong) [inline test/plugins.test.js]: Revert this change.

**Comment by yunong** [comment]
Can you please revert all of the unnecessary styling changes you've made to this repo?

**Comment by yunong** [comment]
Please also add docs about the changes to the audit plugin to docs/index.restdown.

**Comment by leitsubomi** [comment]
Add object check for server object and fixed all style.  Added a draft for storing log data in ring buffer.

File: lib/plugins/logMetrics.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+var bunyan = require('bunyan');
+
+function logMetrics(options) {
```
Comment (DonutEspresso) [inline lib/plugins/logMetrics.js]: Classes should be capitalized.

**Comment by DonutEspresso** [comment]
Echoing Yunong here, lots of styling changes made by jsformat (I'm guessing?) that make the real diff hard to read. Might want to go back with a difftool and revert the styling changes, and we can file another pull request with just the styling changes we want.
